### DISCUSSION TITLE
Manage dotfiles / convert bash script to xonsh

### DISCUSSION THREAD

**Discussion by laguill** [xonsh/xonsh#5707] [Q&A]
Hello,

I would like to better the synthax in xonsh by converting a set of commands from bash to xonsh.
I am following tutoral from [https://www.atlassian.com/git/tutorials/dotfiles](https://www.atlassian.com/git/tutorials/dotfiles) in bash

Here is the bash commands:
```bash
mkdir -p .config-backup
config checkout .bashrc
if [ $? = 0 ]; then
  echo "Checked out config.";
  else
    echo "Backing up pre-existing dot files.";
    config checkout 2>&1 | grep -E "\s+\." | awk {'print $1'} | xargs -I{} mv {} .config-backup/{}
fi;
config checkout
config config status.showUntrackedFiles no
```

I am having this for xonsh:
```xonsh
mkdir -p .config-backup
checkout = !(config checkout)
if check.returncode == 0:
    print("Checked out config.")
else:
    print("Backing up pre-existing dot files.")
    dotfiles = checkout.output.split()
    for dotfile in dotfiles:
        print(dotfile)
config checkout
config config status.showUntrackedFiles no
```

Can you me help convert this command ? 
`    config checkout 2>&1 | grep -E "\s+\." | awk {'print $1'} | xargs -I{} mv {} .config-backup/{}
`

Thanks for your help

**Comment by anki-code** [comment]
This section should help - https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#2-strings-and-arguments-in-shell-commands

**Reply by laguill** [reply]
I made some progress

I am still confused about this `config checkout 2>&1 | grep -E "\s+\." | awk {'print $1'} | xargs -I{} mv {} .config-backup/{}`

> `config checkout` :arrow_right: merge stdout and stderr 
From what I understood in xonsh this the default output  so there is no need to store `config checkout 2>&A` ?

Here is what I have. Is there a more "xonsh" way ?

```xonsh
mkdir -p .config-backup
if (ret := !(config checkout)):
    print("Checked out config.")
else:
    print("Backing up pre-existing dot files.")
    # Store output lines and filter for dotfiles
    dotfiles = [line.split()[1] for line in ret.out.splitlines() if line.split()[1].startswith(".")]
    # Move each dotfile to the backup directory
    for dotfile in dotfiles:
        mv @(dotfile) .config-backup/@(dotfile)
 ```

**Reply by anki-code** [reply]
Following the [xonsh cheatsheet](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#2-strings-and-arguments-in-shell-commands) this is working to me as well:

Bash:
```bash
echo 'hello  .' 2>&1 | grep -E "\s+\." | awk {'print $1'} | xargs -I{} echo {} .config-backup/{}
# hello .config-backup/hello
```
Xonsh (added quotes and using `r`-string):
```xsh
echo 'hello  .' 2>&1 | grep -E r"\s+\." | awk '{print $1}' | xargs '-I{}' echo '{}' '.config-backup/{}'
# hello .config-backup/hello
```

**Reply by laguill** [reply]
The original bash was working.
I just wanted to learn xonsh scripting.
So I tried something easier.

Can you confirmed that 'config checkout' and 'config checkout 2>&1` will produce the same output when using `!()` ?

**Reply by anki-code** [reply]
Looks as expected:
```xsh
!(ls qewqe)
# output='',
# errors="ls: cannot access 'nofile': No such file or directory\n"

!(ls nofile 2>&1)
# output="ls: cannot access 'nofile': No such file or directory"
```

**Comment by laguill** [comment]
Here what I end up with

```xonsh
git clone --bare git@github.com:$GITHUB_USERNAME/myconfig.git $HOME/.dotfiles
function config {
   /usr/bin/git --git-dir=$HOME/.dotfiles/ --work-tree=$HOME $@
}
mkdir -p .config-backup
if (ret := !(config checkout)):
    print("Checked out config.")
else:
    print("Backing up pre-existing dot files.")
    # Store output lines and filter for dotfiles
    dotfiles = [line.split()[1] for line in ret.out.splitlines() if line.split()[1].startswith(".")]
    # Move each dotfile to the backup directory
    for dotfile in dotfiles:
        mv @(dotfile) .config-backup/@(dotfile)
config checkout
config config status.showUntrackedFiles no
```
